### DISCUSSION TITLE
support for Duo Universal Prompt

### DISCUSSION THREAD

**Discussion by fredtj** [dani-garcia/vaultwarden#3953] [General]
Hello,

Vaultwarden's current Duo MFA implementation uses the traditional Duo iframe prompt. Use of this prompt will be deprecated on March 20th, 2024.
To continue using Duo, Vaultwarden needs to migrate to the Duo Universal Prompt before that date.

[https://duo.com/docs/universal-prompt-update-guide](https://duo.com/docs/universal-prompt-update-guide)

Are there any plans to update the Duo implementation in Vaultwarden before March 20th 2024?

Thanks in advance!

**Comment by BlackDex** [comment]
I think this has to be done by Bitwarden since that is build-in into the web-vault and they maintain and develop the web-vault.

**Reply by misilot** [reply]
It was recently added by Bitwarden

https://community.bitwarden.com/t/support-for-duo-universal-prompt/37017/22
Release notes: https://bitwarden.com/help/releasenotes/#2024-2-3

**Comment by fredtj** [comment]
It seems already available in Bitwarden via SSO, see https://duo.com/docs/sso-bitwarden and https://bitwarden.com/help/about-sso/

Vaultwarden as far as I can tell does not yet have this option.

Regards

**Comment by Gerardv514** [comment]
I have checked my duo account and I see universal prompt is selected. Are we supposed to do something else, I am still seeing the traditional prompt, should I be seeing the universal prompt yet?

**Reply by agentdr8** [reply]
I had to create another Duo app and select Bitwarden in the Duo admin console before the universal prompt showed as being available. Then changed the global Duo keys in the VW admin panel. But after doing that, I'm still seeing an error trying to use the Universal Prompt with the testing docker image:

![image](https://github.com/dani-garcia/vaultwarden/assets/3790868/5b4b920c-75bd-4cf6-86e8-34218844c010)

And after seeing that error, now the new BW app in Duo is no longer showing the option for Universal Prompt, so it seems like the integration in VW web 2024.3.1 is still not compatible with UP.

**Reply by Gerardv514** [reply]
I now see in Duo Admin portal, when you create another DUO app for Bitwarden it now says DUO Universal backend has been update.

I believe we now have to wait for VW to make the necessary updates to support the Universal Prompt; that is probably why you're receiving the setup error.

**Reply by agentdr8** [reply]
@BlackDex - If BW implemented the new Duo Universal Prompt changes in 2024.2.3, are those changes not part of VW's web-vault 2024.3.1?

**Reply by misilot** [reply]
Server Side pieces would need to be updated as well per https://github.com/dani-garcia/bw_web_builds/pull/157#issuecomment-2020897313

**Reply by EarSum** [reply]
I still have the same issue at 2024/7/16. But the duo admin plannel shows "Last Date of Support for the Traditional Prompt: September 30, 2024", so the fix might be a little urgent for duo user. In case you can't log in at that time, it might be a good idea to use one of the other two-step authentication methods like email.

**Comment by azaloum90** [comment]
Any Update on this?  Not sure what is entailed from the server side?

**Reply by BlackDex** [reply]
There need to be server side changes for this. Which are not implemented.

**Reply by mjd2000** [reply]
Shouldn't DUO support be removed from the main page then? It does not seem to be working at all without the upgrade to universal prompt

**Comment by azaloum90** [comment]
It does work without the universal prompt, but it could stop working at any time which is obviously problematic.

Adam

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by Simon-CR** [comment]
also looking forward to this...

**Comment by obasigeorge** [comment]
It has actually stopped working.

**Reply by JH220** [reply]
> It has actually stopped working.

It still works for me, but when I try to log in via the Windows Desktop app, I can't scroll down or make the window bigger, so unfortunately I can no longer log in there and have to do it via a different method.
The problem is that it can stop working at any time, the Universal Prompt is already in the official instance, so it should also be possible to do this via Vaultwarden.

**Comment by airocle** [comment]
This is fairly critical, what's the status?

**Reply by Gerardv514** [reply]
It’s being worked on in a pull. I have and use duo too, but I also have ability to use  Authenticator app. Are you able to use that in the mean time?

**Reply by airocle** [reply]
> It’s being worked on in a pull. I have and use duo too, but I also have ability to use Authenticator app. Are you able to use that in the mean time?

We use DUO in our org, I'd rather not have to get everyone to manually configure another authenticator app as a stop-gap if possible.

**Reply by airocle** [reply]
Very glad to hear it's being worked on, what's the pull request in question?

**Reply by BlackDex** [reply]
It's already merged and available in the `testing` tagged container images.

**Reply by MMagTech** [reply]
Just updated to the testing tag and all is working.  Dropping a comment to track completion to the update being in the latest tag.

**Reply by JamieWhite** [reply]
@BlackDex I've just been testing this and the redirect doesn't seem like it works for installations with custom paths. For example my bitwarden is installed at: `https://website.net/path` 

when I login via duo it forwards me to: https://website.net/duo-redirect-connector.html?client=web&state=xxx&code=yyy

This should be: https://website.net/path/duo-redirect-connector.html?client=web&state=xxx&code=yyy

If I update this manually in my browser it says im logged in and can close the window. When I close the window it works fine. 

I think the code just needs to be updated to use `Domain URL` from /admin/ where I've got my path input correctly.

**Reply by JamieWhite** [reply]
Managed to hack this with an nginx redirect for now:

```
location = /duo-redirect-connector.html {
  return 301 /path/duo-redirect-connector.html?$args;
}
```

**Reply by BlackDex** [reply]
Thanks for providing the workaround and creating the issue :)
